From 559725c1ee4103b05845c170f6e0e2dab2631f93 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 13:12:37 -0400
Subject: [ethernet] i40evf: don't stop watchdog if it hasn't started

Message-id: <1409663591-32531-115-git-send-email-sassmann@redhat.com>
Patchwork-id: 90586
O-Subject: [RHEL7.1 PATCH 114/148] i40evf: don't stop watchdog if it hasn't started
Bugzilla: 1091127
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Mitch Williams <mitch.a.williams@intel.com>

If the VF driver fails to complete early init, then rmmod can cause a
softlock when the driver tries to stop a watchdog timer that never even
got initialized.

Add a check to see if the timer is actually initialized before stopping
it.

Change-ID: Id9d550aa8838e07f4b02afe7bc017ef983779efc
Signed-off-by: Mitch Williams <mitch.a.williams@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit e5d17c3ed2cf0485c6b85d783e1d20a8cfad93c1)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40evf/i40evf_main.c b/drivers/net/ethernet/intel/i40evf/i40evf_main.c
index b473172..36b1ad7 100644
--- a/drivers/net/ethernet/intel/i40evf/i40evf_main.c
+++ b/drivers/net/ethernet/intel/i40evf/i40evf_main.c
@@ -2408,7 +2408,9 @@ static void i40evf_remove(struct pci_dev *pdev)
   i40evf_reset_interrupt_capability(adapter);
  }
 
- del_timer_sync(&adapter->watchdog_timer);
+ if (adapter->watchdog_timer.function)
+  del_timer_sync(&adapter->watchdog_timer);
+
  flush_scheduled_work();
 
  if (hw->aq.asq.count)
-- 
1.7.1